lib/cext/preprocess.rb

```diff
@@ -114,12 +114,25 @@ def self.patch(file, contents, directory)
     contents
   end
 
+  require 'stringio'
+
   if __FILE__ == $0
-    puts "#line 1 \"#{ARGF.filename}\""
+    out = StringIO.new
 
-    contents = patch(ARGF.filename, File.read(ARGF.filename), Dir.pwd)
+    original = File.read(ARGF.filename)
+    contents = patch(ARGF.filename, original.clone, Dir.pwd)
     contents.each_line do |line|
-      puts preprocess(line)
+      out.puts preprocess(line)
+    end
+
+    out.rewind
+    output = out.read
+    $stdout.print "#line 1 \"#{ARGF.filename}\"\n" + output
+
+    if ENV['PREPROCESS_DEBUG'] && original != output
+      patched_file_name = '.' + ARGF.filename.gsub(/\.c$/, '.patched.c')
```

Comment:
`sub` not `gsub`, but `File.basename(file, ".*")` + File.dirname would be cleaner.